RFC-0009-jdbc-join-push-down.md

```diff
@@ -0,0 +1,798 @@
+
+#### 3. Use MultiJoinNode to group Jdbc Tables based on connector name
+- 3.1. We take each item of multiJoinNode's sourceList and check if it’s a connector which supports join push down. For this we have introduced a new capability in ConnectorCapabilities named "SUPPORTS_JOIN_PUSHDOWN”.
+```
+public enum ConnectorCapabilities
+{
+    NOT_NULL_COLUMN_CONSTRAINT,
+    SUPPORTS_REWINDABLE_SPLIT_SOURCE,
+    SUPPORTS_PAGE_SINK_COMMIT,
+    SUPPORTS_JOIN_PUSHDOWN
+}
+```
+- 3.2. In JdbcConnector, we set this capability to enable Join Pushdown. So that all Jdbc connectors will get this join pushdown capability. 
+```
+@Override
+public Set<ConnectorCapabilities> getCapabilities()
+{
+ return immutableEnumSet(NOT_NULL_COLUMN_CONSTRAINT, SUPPORTS_JOIN_PUSHDOWN);
+}
+```
+- 3.3. Once it identifies the connector as pushdown supported, it creates a Map with key as connector name and value as a List of tables which are from the connector. [Details available here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#3-use-multijoinnode-to-group-jdbc-tables-based-on-connector-name-1)
+- 3.4. This ensures that no other connector is affected by this optimizer. Only connectors with Join pushdown capability will be pushed down.
+- In-depth details are available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#3-use-multijoinnode-to-group-jdbc-tables-based-on-connector-name-1)
+
+#### 4. Grouping tables for creating join query - based on [JDBC datasource capability](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#join-query-pushdown-in-presto-jdbc-datasource)
+- 4.1. JoinTables (List of ConnectorTableHandle) creation happens from the Map which is created above. [Point number 3.3]
+- 4.2. For each item in map, based on connector, we get a list of tables/nodes. Each node is then analyzed for join pushdown capability and either added to JoinTables List or added back to rewrittenList (If it can not be pushed down).
+- In-depth details are available [here](https://github.com/Thanzeel-Hassan-IBM/rfcs/blob/main/RFC-0009-jdbc-join-push-down.md#4-build-join-relation-for-the-grouped-tables-from-all-the-join-predicates)
+
+#### 5. If we are able to create a JoinTables list, then we create a single table scan for that and then add to the rewrittenList.
+- 5.1. If there are 4 tables in JoinTables list against Postgres, then we create a single table scan node with ConnectorHandleSet 
+- 5.2. Inside the ConnectorHandleSet, these 4 tables will be there.
+- 5.3. This rewrittenList is used to create another multiJoinNode (rewrittenMultiJoinNode).
+
+#### 6. Create a joinNode for each sourceList
+- Iterate over the rewrittenMultiJoinNode, for each sourceList, call createLeftDeepJoinTree() method. This creates a joinNode with all the nodes in the sourceList.
```

Comment:
add link to `Recreate left deep join node from the MultiJoinNode source list `for the details

Reply:
added